lib/fluent/plugin/sd_srv.rb

```diff
@@ -0,0 +1,130 @@
+#
+# Fluentd
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+require 'resolv'
+
+require 'fluent/plugin_helper'
+require 'fluent/plugin/service_discovery'
+
+module Fluent
+  module Plugin
+    class SrvServiceDiscovery < ServiceDiscovery
+      include PluginHelper::Mixin
+
+      Plugin.register_sd('srv', self)
+
+      helpers :timer
+
+      config_param :service, :string
+      config_param :proto, :string, default: 'tcp'
+      config_param :hostname, :string
+      config_param :dns_server_host, :string, default: nil
+      config_param :interval, :integer, default: 60
+      config_param :dns_lookup, :bool, default: true
+
+      def initialize
+        super
+        @target = nil
+      end
+
+      def configure(conf)
+        super
+
+        @target = "_#{@service}._#{@proto}.#{@hostname}"
+        @dns_resolve =
+          if @dns_server_host.nil?
+            Resolv::DNS.new
+          elsif @dns_server_host.include?(':') # e.g. 127.0.0.1:8600
+            host, port = @dns_server_host.split(':', 2)
+            Resolv::DNS.new(nameserver_port: [[host, port.to_i]])
+          else
+            Resolv::DNS.new(nameserver: @dns_server_host)
+          end
+
+        @services = fetch_srv_record
+      end
+
+      def start(queue)
+        timer_execute(:"sd_srv_record_#{@target}", @interval) do
+          refresh_srv_records(queue)
+        end
+
+        super()
+      end
+
+      private
+
+      def refresh_srv_records(queue)
+        s = begin
+              fetch_srv_record
+            rescue => e
+              @log.error("sd_srv: #{e}")
+              return
+            end
+
+        if s.nil? || s.empty?
+          return
+        end
+
+        diff = []
+        join = s - @services
+        # Need service_in first to guarantee that server exist at least one all time.
+        join.each do |j|
+          diff << ServiceDiscovery.service_in_msg(j)
+        end
+
+        drain = @services - s
+        drain.each do |d|
+          diff << ServiceDiscovery.service_out_msg(d)
+        end
+
+        @services = s
+
+        diff.each do |a|
+          queue.push(a)
+        end
+      end
+
+      def fetch_srv_record
+        adders =
+          begin
+            @dns_resolve.getresources(@target, Resolv::DNS::Resource::IN::SRV)
+          rescue => e
+            raise e               # TODO
```

Comment:
What is this TODO?

Reply:
Never mind.  I thought it needed to rescue and output something log. but it's done in refresh_srv_records.

https://github.com/fluent/fluentd/pull/2876/commits/e7d82e05c311e381709100e4a6c88d080f96d675